vpr/src/device/rr_graph/rr_graph_obj.cpp

```diff
@@ -0,0 +1,1753 @@
+            ++iedge;
+        }
+    }
+
+    return iedge;
+}
+
+/* 
+ * Get the number of configurable output edges of a node
+ * TODO: we would use the node_num_configurable_out_edges() 
+ * when the rr_graph edges have been partitioned 
+ * This can avoid unneccessary walkthrough
+ */
+short RRGraph::node_num_configurable_out_edges(RRNodeId node) const {
+    /* Ensure a valid node id */
+    VTR_ASSERT_SAFE(valid_node_id(node));
+
+    /* Each all the edges ends at this node */
+    short iedge = 0;
+    for (auto edge : node_out_edges(node)) {
+        if (true == edge_is_configurable(edge)) {
+            ++iedge;
+        }
+    }
+
+    return iedge;
+}
+
+/* 
+ * Get the number of non-configurable input edges of a node
+ * TODO: we would use the node_num_configurable_in_edges() 
+ * when the rr_graph edges have been partitioned 
+ * This can avoid unneccessary walkthrough
+ */
+short RRGraph::node_num_non_configurable_in_edges(RRNodeId node) const {
+    /* Ensure a valid node id */
+    VTR_ASSERT_SAFE(valid_node_id(node));
+
+    /* Each all the edges ends at this node */
+    short iedge = 0;
+    for (auto edge : node_in_edges(node)) {
+        if (true == edge_is_non_configurable(edge)) {
+            ++iedge;
+        }
+    }
+
+    return iedge;
+}
+
+/* 
+ * Get the number of non-configurable output edges of a node
+ * TODO: we would use the node_num_configurable_out_edges() 
+ * when the rr_graph edges have been partitioned 
+ * This can avoid unneccessary walkthrough
+ */
+short RRGraph::node_num_non_configurable_out_edges(RRNodeId node) const {
+    /* Ensure a valid node id */
+    VTR_ASSERT_SAFE(valid_node_id(node));
+
+    /* Each all the edges ends at this node */
+    short iedge = 0;
+    for (auto edge : node_out_edges(node)) {
+        if (true == edge_is_non_configurable(edge)) {
+            ++iedge;
+        }
+    }
+
+    return iedge;
+}
+
+/* Get the edge id that connects two nodes */
+RREdgeId RRGraph::get_node_to_node_edge(RRNodeId from_node, RRNodeId to_node) const {
+    /* Make sure we will access a valid node */
+    VTR_ASSERT_SAFE(valid_node_id(from_node));
+    VTR_ASSERT_SAFE(valid_node_id(to_node));
+
+    /* check the output edges of from_node  */
+    for (auto edge : node_out_edges(from_node)) {
+        /* Check if the sink node of the edge is the to_node */
+        if (to_node == edge_sink_node(edge)) {
+            return edge; /* we find it return here */
+        }
+    }
+    /* We find nothing, return an RRGRAPH_OPEN ID */
+    return RRGRAPH_OPEN_EDGE_ID;
+}
+
+/* Get the switch id that connects two nodes */
+RRSwitchId RRGraph::get_node_to_node_switch(RRNodeId from_node, RRNodeId to_node) const {
+    /* Make sure we will access a valid node */
+    VTR_ASSERT_SAFE(valid_node_id(from_node));
+    VTR_ASSERT_SAFE(valid_node_id(to_node));
+
+    RREdgeId edge = get_node_to_node_edge(from_node, to_node);
+    if (RRGRAPH_OPEN_EDGE_ID == edge) {
```

Comment:
if (!edge)

Reply:
Done. I use INVALID() just be consistent as how it is used in other functions.